### PR TITLE
feature: enable all2all communication for MoE model

### PR DESCRIPTION
**Pull request by rebel-jangys**
### 🚀 Summary of Changes

> *What does this PR do? What feature, fix, or improvement does it bring?*

---

### 📌 Related Issues / Tickets

* Resolves #
* Related to #

---

### ✅ Type of Change

* [ ] 🚀 Release (`release`)
* [ ] ✨ Feature (`feature`)
* [ ] 🧠 Model support (`model`)
* [ ] 🧬 Core engine changes (`core`)
* [ ] 🛠 Bug fix (`fix`)
* [ ] ⚙️ Performance improvement (`perf`)
* [ ] 🔁 Refactor or code cleanup (`refactor`)
* [ ] 📄 Documentation (`docs`)
* [ ] ❓ Other (`other`): please describe

---

### 🧪 How to Test

1. Run `...`
2. Verify output: `...`
3. Edge case tested: `...`

---

### 📸 Screenshots / Logs (if applicable)

---

### 📋 Checklist

* [ ] PR title follows Conventional Commits format
* [ ] This PR is linked to an existing issue
* [ ] The test method is described, and the expected result is clearly stated
* [ ] Relevant documentation has been updated (if applicable)

---

### 💬 Notes

---

### CODE REVIEW

File: vllm_rbln/model_executor/layers/fused_moe/layer.py
Comment (rebel-ykchoi) [inline vllm_rbln/model_executor/layers/fused_moe/layer.py]: I think it is dead codes because here is --- DP == 1 path ---.

File: vllm_rbln/rbln_envs.py [hunk 1]
```diff
@@ -46,6 +45,9 @@
     VLLM_RBLN_DECODE_BATCH_BUCKET_MANUAL_BUCKETS: list[int] = []
     VLLM_RBLN_USE_CUSTOM_KERNEL: bool = False
     VLLM_RBLN_AUTO_PORT: bool = True
+    VLLM_RBLN_CCL_ALL2ALL_P2P: bool = False
```
Comment (rebel-ykchoi) [inline vllm_rbln/rbln_envs.py]: `VLLM_RBLN_CCL_ALL2ALL_P2P` is not used anywhere.
Is this a placeholder for later use?

File: vllm_rbln/model_executor/layers/fused_moe/layer.py
Comment (rebel-ykchoi) [inline vllm_rbln/model_executor/layers/fused_moe/layer.py]: `unquantized_fused_moe_method_custom` changed a lot, so we can’t safely keep the old `unquantized_fused_moe_method_rbln` fallback anymore. It can easily break because it expects different routing inputs.

I suggest deleting both the `VLLM_RBLN_MOE_CUSTOM_KERNEL` env var and the`unquantized_fused_moe_method_rbln` implementation, and always using the custom MoE path. This keeps one consistent behavior and reduces future bugs and maintenance.

File: vllm_rbln/model_executor/layers/fused_moe/layer.py [hunk 2]
```diff
@@ -438,27 +440,411 @@ def fused_moe_forward_rbln(
...
+                    topk_weights = topk_weights / topk_weights.sum(
+                        dim=0, keepdim=True
+                    ).clamp_min(1e-20)
+                masked_routing_weights = torch.zeros_like(
+                    all_router_logits_t
+                )  # [E, R*max_pad]
+                masked_routing_weights.scatter_(0, selected_experts, topk_weights)
+
+        # Apply token mask to zero out padded positions per DP rank
+        use_moe_tokens_mask = envs.VLLM_RBLN_USE_MOE_TOKENS_MASK
```
Comment (rebel-ykchoi) [inline vllm_rbln/model_executor/layers/fused_moe/layer.py:419]: I think
```
        masked_routing_weights_depadded = masked_routing_weights
```
is required for not use_moe_tokens_mask case.

File: vllm_rbln/model_executor/layers/fused_moe/layer.py [hunk 3]
```diff
@@ -438,27 +440,411 @@ def fused_moe_forward_rbln(
...
+        # [E, R*max_pad] for dim=0 topk (select top_k experts per token)
+        all_router_logits_t = all_router_logits.transpose(0, 1)  # [E, R*max_pad]
+
+        scoring_func = getattr(self, "scoring_func", "softmax")
+        e_score_correction_bias = getattr(self, "e_score_correction_bias", None)
+        use_grouped_topk = getattr(self, "use_grouped_topk", False)
+
+        if scoring_func == "sigmoid":
+            if use_grouped_topk:
+                masked_routing_weights = _apply_grouped_topk_torch(
```
Comment (rebel-ykchoi) [inline vllm_rbln/model_executor/layers/fused_moe/layer.py:335]: In `_apply_grouped_topk_torch` function "sigmoid" is not supported for scoring_func.
I think it might cause silent error cases for some models.

**Review by rebel-ykchoi — CHANGES_REQUESTED** [review]

File: vllm_rbln/model_executor/layers/fused_moe/layer.py [hunk 4]
```diff
@@ -438,27 +440,411 @@ def fused_moe_forward_rbln(
...
+        masked_routing_weights_depadded = F.pad(
+            masked_routing_weights, (0, -pad_size), value=0.0
+        )
+
+    # pass as [t, E] to quant_method.apply (it will be reshaped inside)
     final_hidden_states = self.quant_method.apply(
         layer=self,
         x=hidden_states,
-        router_logits=router_logits,
+        router_logits=masked_routing_weights,
```
Comment (rebel-ykchoi) [inline vllm_rbln/model_executor/layers/fused_moe/layer.py:654]: Is it correct to apply `masked_routing_weights` instead of `masked_routing_weights_depadded`?

**Review by rebel-ykchoi — CHANGES_REQUESTED** [review]

File: vllm_rbln/model_executor/layers/fused_moe/layer.py (hunk 3 above)
Comment (rebel-jangys) [reply vllm_rbln/model_executor/layers/fused_moe/layer.py:335]: @rebel-hishin 이거 말씀하신 부분 수정 필요한 것 같은데 확인해주실 수 있나요?

File: vllm_rbln/model_executor/layers/fused_moe/layer.py
Comment (rebel-jangys) [reply vllm_rbln/model_executor/layers/fused_moe/layer.py]: fixed in [256719a](https://github.com/RBLN-SW/vllm-rbln/pull/605/commits/256719a2740aa30f8a31a2b3b2ddbbe8868941a3)

File: vllm_rbln/rbln_envs.py (hunk 1 above)
Comment (rebel-jangys) [reply vllm_rbln/rbln_envs.py]: fixed in [60b744e](https://github.com/RBLN-SW/vllm-rbln/pull/605/commits/60b744e27e31d184aadf78077bdc961cf477c232)

File: vllm_rbln/model_executor/layers/fused_moe/layer.py
Comment (rebel-jangys) [reply vllm_rbln/model_executor/layers/fused_moe/layer.py]: okay. I deprecated default torch impl for moe op in commit [18b945b](https://github.com/RBLN-SW/vllm-rbln/pull/605/commits/18b945b777c506e0e3ec2b2b56c5ae81fd86f1d8)

File: vllm_rbln/model_executor/layers/fused_moe/layer.py (hunk 2 above)
Comment (rebel-jangys) [reply vllm_rbln/model_executor/layers/fused_moe/layer.py:419]: fixed in [da9a891](https://github.com/RBLN-SW/vllm-rbln/pull/605/commits/da9a891ec581d8ff08d974e62218ee02781fd3fb)

File: vllm_rbln/model_executor/layers/fused_moe/layer.py (hunk 4 above)
Comment (rebel-jangys) [reply vllm_rbln/model_executor/layers/fused_moe/layer.py:654]: Thanks, you're right. Fixed in [da9a891](https://github.com/RBLN-SW/vllm-rbln/pull/605/commits/da9a891ec581d8ff08d974e62218ee02781fd3fb)

**Review by rebel-hishin — APPROVED** [review]

File: vllm_rbln/model_executor/layers/fused_moe/layer.py [hunk 5]
```diff
@@ -438,54 +344,411 @@ def fused_moe_forward_rbln(
...
+
+            # unpacked: [R, max_pad, H] → flatten to [R*max_pad, H] for MoE
+            gathered_hidden = unpacked.reshape(R * max_pad, H_dim)
+        # else: gathered_hidden was already computed via all-gather before router
+
+        # --- Step 5: MoE FFN computation ---
+        final_hidden_states = self.quant_method.apply(
+            layer=self,
+            x=gathered_hidden,
+            router_logits=masked_routing_weights,  # [E*max_pad, T_padded]
```
Comment (rebel-ykchoi) [inline vllm_rbln/model_executor/layers/fused_moe/layer.py:480]: For the DP path (dp_size > 1), quant_method.apply is called with masked_routing_weights (which may include 64-align padding), while the DP==1 path passes masked_routing_weights_depadded ([E, t]).

Is that intentional? If so, could you briefly explain why the padded tensor is needed for the DP MoE/compile path?

File: vllm_rbln/model_executor/layers/fused_moe/layer.py [hunk 6]
```diff
@@ -260,6 +124,121 @@ def get_tokens_mask(num_tokens: int, left=1.0, right=0.0, device=None):
     return tokens_mask
 
 
+def get_tokens_mask_DP(num_tokens: int, dp_rank: int, left=1.0, right=0.0):
```
Comment (rebel-ykchoi) [inline vllm_rbln/model_executor/layers/fused_moe/layer.py]: There are some dead codes remain.
```
get_tokens_mask_DP
get_masked_routing_weights
fused_moe_naive_multicast_rbln
```

**Review by rebel-ykchoi — CHANGES_REQUESTED** [review]

File: vllm_rbln/model_executor/layers/fused_moe/layer.py (hunk 6 above)
Comment (rebel-jangys) [reply vllm_rbln/model_executor/layers/fused_moe/layer.py]: Okay, I cleaned up.

File: vllm_rbln/model_executor/layers/fused_moe/layer.py (hunk 5 above)
Comment (rebel-jangys) [reply vllm_rbln/model_executor/layers/fused_moe/layer.py:480]: Thanks! I removed an explicit padding.

**Review by rebel-ykchoi — APPROVED** [review]
Thanks a lot for your effort!!!

File: vllm_rbln/model_executor/layers/fused_moe/all2all.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+def _ccl_combine_receive_fake(
+    recv_buffer: Tensor,
+    router_logits: Tensor,
+    expert_map: Tensor,
+    hidden_states: Tensor,
+    rank_id: int,
+) -> Tensor:
+    t_dim = recv_buffer.shape[1]
+    H_dim = recv_buffer.shape[2]
+    return torch.empty(t_dim, H_dim, dtype=recv_buffer.dtype)
```
Comment (rebel-jongho) [inline vllm_rbln/model_executor/layers/fused_moe/all2all.py:322]: how about move this code to rebel-compiler?
